perceval/rendering/circuit/renderer.py

```diff
@@ -484,114 +575,217 @@ def _add_shape(self, lines, circuit, content, w, shape_fn=None, pos=None):
-                        self._out_port_pos[ppos.initial_mode].y = ppos.y
-                    else:
-                        self._mode_style[i] = ModeStyle.PHOTONIC
-
-                else:
-                    ppos.y = circuit.perm_vector[ppos.y - lines[0]] + lines[0]
-                    ppos.fixed = None
-
-        # Out heralds
+    def _update_mode_style(
+            self, lines, circuit, w: int, subc_mode: bool = False):
         if not isinstance(circuit, PERM) and not subc_mode:
-            herald_out_mode, component_out_mode = self._search_component_in_out_herald_info(circuit)
-            if herald_out_mode is not None:
-                self._mode_style[lines[0] + component_out_mode] = ModeStyle.HERALD
-                self._out_port_pos[herald_out_mode].y = lines[0] + component_out_mode
-                self._out_port_pos[herald_out_mode].x = self._chart[lines[0] + component_out_mode] + w
-
+            input_heralds = {}
+            output_heralds = {}
+
+            herald_info = self._herald_info if self._herald_info else {}
+            if circuit in herald_info:
+                output_heralds = herald_info[circuit].output_heralds
+                input_heralds = herald_info[circuit].input_heralds
+
+            # Position input and output heralds
+            for in_mode, herald_in_mode in input_heralds.items():
+                self._in_port_pos[herald_in_mode].y = lines[0] + in_mode
+                self._in_port_pos[herald_in_mode].x = \
+                    self._chart[lines[0] + in_mode]
+                # Start drawing this mode in "photonic" style
+                self._mode_style[lines[0] + in_mode] = ModeStyle.PHOTONIC
+            for out_mode, herald_out_mode in output_heralds.items():
+                self._out_port_pos[herald_out_mode].y = lines[0] + out_mode
+                self._out_port_pos[herald_out_mode].x = \
+                    self._chart[lines[0] + out_mode] + w
+                # Stop drawing this mode (set it in "herald" style)
+                self._mode_style[lines[0] + out_mode] = ModeStyle.HERALD
         if isinstance(circuit, PERM):
+            m0 = lines[0]
             out_modes = copy.copy(self._mode_style)
             for m_input, m_output in enumerate(circuit.perm_vector):
-                out_modes[m_output+m0] = self._mode_style[m_input+m0]
+                out_modes[m_output + lines[0]] = self._mode_style[m_input + m0]
             self._mode_style = out_modes
 
-        for p in self._in_port_pos:
-            if p.fixed is None:
-                p.fixed = False
-        # END Mode tracking + herald positionning algo
-
     def append_circuit(self, lines, circuit, content, pos=None):
         w = self._skin.get_width(circuit)
-        self._add_shape(lines, circuit, content, w, pos=pos)
-        self._update_mode_style(lines, circuit, w)
-        for i in range(lines[0], lines[-1] + 1):
-            self._chart[i] += w
+        if w:
```

Comment:
Was this added in order to sometimes use `Barrier` with 0 width?
Anyway, nice add.

Reply:
Yes, this was added for my initial implementation that had both "Identity" (acts like a barrier but invisible on the circuit), and "Barrier".

I don't think that's needed anymore, but if you think there might be a use case where some circuit elements have width=0, I can make sure it's used everywhere needed.